Add FreeBSD operations & facts

DtxdF opened the pull request.

DtxdF committed:
Add FreeBSD operations & facts

DtxdF commented:
fixed!

DtxdF committed:
Escape more strings

DtxdF committed:
Fix logic in SRV_CUSTOM

DtxdF committed:
Fix type-checking errors

Fizzadar commented on pyinfra/operations/freebsd/__init__.py:
```diff
@@ -0,0 +1,12 @@
+# This file only exists to support:
```
I really like the use of a submodule for all the `freebsd` operations, neat.

Fizzadar reviewed (commented):
Hi @DtxdF! Thank you for the PR, this looks great! Would it be possible to add some tests for the operations? (plenty of examples here: https://github.com/pyinfra-dev/pyinfra/tree/3.x/tests/operations).

DtxdF commented:
> Hi @DtxdF! Thank you for the PR, this looks great! Would it be possible to add some tests for the operations? (plenty of examples here: https://github.com/pyinfra-dev/pyinfra/tree/3.x/tests/operations).

Finally I have a time.

I have seen the logic in both `tests/operations` and `tests/facts` but how can I test them on my system?

Fizzadar commented:
> I have seen the logic in both `tests/operations` and `tests/facts` but how can I test them on my system?

There's a script `scripts/dev-test.sh` that runs them or you can use `pytest` directly. The [contributing guide](https://docs.pyinfra.com/en/3.x/contributing.html) has complete instructions!

0EVSG commented:
Nice to see this. @DtxdF, contact me if you're interested in getting this into the FreeBSD port early, or maybe even taking maintainership of the port - I'm currently lacking time to invest more into side projects like this.

DtxdF committed:
Add tests for FreeBSD operations & facts

* Change `tests/test_operations.py` to accept freebsd operations
  - Fix `ValueError: too many values to unpack (expected 2)`
* Simplify fact `freebsd.ServiceScript`.
* Add missing parameter `jail` in `freebsd.pkg` operation.
* Change Enum to simple "constants" in `freebsd.sysrc` and `freebsd.service`
  to use strings instead of the properties to make the tests work and to use
  those operations from the CLI.
* Change the `state` argument to `srvstate` in `freebsd.service.service`
  to avoid conflicts with `pyinfra_cli/util.py:load_func` when invoking
  the operation from the CLI.

DtxdF committed:
Remove unused `enum` library

DtxdF committed:
Remove conflicting blank line

DtxdF commented:
> Nice to see this. @DtxdF, contact me if you're interested in getting this into the FreeBSD port early, or maybe even taking maintainership of the port - I'm currently lacking time to invest more into side projects like this.

Great! I will send you a patch in Bugzilla! Thanks!

DtxdF commented:
> > I have seen the logic in both `tests/operations` and `tests/facts` but how can I test them on my system?
> 
> There's a script `scripts/dev-test.sh` that runs them or you can use `pytest` directly. The [contributing guide](https://docs.pyinfra.com/en/3.x/contributing.html) has complete instructions!

Thank you! The tests are already with some changes, let me know if everything goes well!

DtxdF commented:
@Fizzadar 

Are you ok with this PR or do I need to do something else? I am very anxious to migrate from Ansible to Pyinfra :D

Fizzadar reviewed (approved):
@DtxdF apologies for the long delay here - thank you for adding the tests and cleanup, looks good to me!

Test failure unrelated, merging!
